ci: update run_easymesh_cert.sh to support upload to owncloud (Patch 2)

tomereli opened the pull request:
Add a script for uploading file[s] / folders to owncloud via curl.
(based on https://doc.owncloud.com/server/user_manual/files/access_webdav.html#accessing-files-using-curl) and update `run_easymesh_cert.sh` to support uploading to owncloud.

Example logs uploaded by the script:
https://ftp.essensium.com/owncloud/index.php/apps/files/?dir=/Nightly/agent_certification/2020-03-25_19-19-40&fileid=712397

Commit:
ci: add script for uploading to owncloud via curl

Add a script for uploading file[s] / folders to owncloud via curl.
This has been developed based on
https://doc.owncloud.com/server/user_manual/files/access_webdav.html#accessing-files-using-curl.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert.sh: add upload support

Add upload to owncloud support to run_easymesh_cert.sh, using the
upload_to_owncloud.sh script.
Before uploading, move the downloaded prplmesh.ipk and
the prplmesh.buildinfo files into the top level log diretory,
so we have it in the cloud.
While we're at it, remove the calculation and prints of pass/fail since
its already done in run_test_file.sh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: download_ipk.sh rename buildinfo file

Since the prplmesh.buildinfo file may be uploaded to owncloud, it is
much easier to have it as regular text file so we can view it from the
web interface. Therefore, update the download_ipk.sh to rename
prplmesh.buildinfo into prplmesh_buildinfo.txt when downloading.

Update the new file name (and some more minor cleanups) in
run_easymesh_cert.sh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert.sh don't use run for run_test_file.sh

The run implementation exists on error, and since we want to upload
logs (especially) when there are failures, we don't want to exit the
script if it happens.
Remove use of 'run' when calling run_test_file.sh script.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+scriptdir="$(cd "${0%/*}" && pwd)"
+rootdir=$(realpath "$scriptdir/../")
+
+# shellcheck source=functions.sh
```
Does this work? I'd expect it should be
```suggestion
# shellcheck source=../tools/functions.sh
```

tomereli replied:
It doesn't work either way, but I'll change to your suggestion

rmelotte commented on tools/download_ipk.sh:
```diff
@@ -22,12 +23,12 @@ download() {
 
 main() {
-    if ! OPTS=$(getopt -o 'h' --long help,branch:,device:,path: -n 'parse-options' -- "$@"); then
+    if ! OPTS=$(getopt -o 'hq' --long help,quite,branch:,device:,path: -n 'parse-options' -- "$@"); then
```
From the parsing and the help, 'q' should be 'v' and 'quite' should be 'verbose'.

arnout replied:
```suggestion
    if ! OPTS=$(getopt -o 'hv' --long help,verbose,branch:,device:,path: -n 'parse-options' -- "$@"); then
```

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+. "$rootdir/tools/functions.sh"
+
+usage() {
+    echo "usage: $(basename "$0") [-hv] <user> [user] <password> [password] <remote-path> [path] <local-path> [path]"
```
I read this as: you can call it as `upload_to_owncloud.sh arnout arnoutpassword path/on/server ./file` or as `upload_to_owncloud.sh arnout user arnoutpassword password path/on/server path ./file path` - which makes no sense.

In other words:
```suggestion
    echo "usage: $(basename "$0") [-hv] <user> <password> <remote-path> <local-path>"
```

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+    while true; do
+        case "$1" in
+            -h | --help)            usage; exit 0;;
+            -v | --verbose)         export VERBOSE=true; REDIRECT=">/dev/null 2>&1"; shift;;
```
Shouldn't this be
```suggestion
            -v | --verbose)         export VERBOSE=true; REDIRECT=""; shift;;
```

and set REDIRECT as a default?

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+    [ -n "$1" ] || { usage; err "Missing local-path"; exit 1; }
+    local_path="$1"; shift
+
+    info "upload $local_path to $OWNCLOUD_URL/$remote_path user $user password $password"
```
I wouldn't echo the password - ever.

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+    echo "      password - owncloud password"
+    echo "      remote-path - path in the cloud, relative to the user home to upload"
+    echo "      local-path - local file or folder to upload"
+    echo ""
```
```suggestion
    echo "usage: $(basename "$0") [-hv] <remote-path> <local-path>"
    echo "  options:"
    echo "      -h|--help - display this help."
    echo "      -v|--verbose - enable verbosity"
    echo "      -u|--url - webDAV URL (default https://ftp.essensium.com/owncloud/remote.php/dav/files/USERNAME)"
    echo
    echo "  positional arguments:"
    echo "      remote-path - path in the cloud, relative to the user home to upload"
    echo "      local-path - local file or folder to upload"
    echo ""
    echo "  Credentials are taken from $HOME/.netrc"
    echo "  Make sure it has a line like this:"
    echo "    machine ftp.essensium.com login <user> password <password>"
    echo ""
```

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+            while read -r -s dir; do
+                printf . # show progress
+                dbg "Create directory: $dir" 
+                curl -u "$user":"$password" -X MKCOL "$OWNCLOUD_URL/$user/$remote_path/$dir" "$REDIRECT"
```
Passwords on the command line are a security nightmare. True, it's going to use an account with very limited permissions, but still we don't like exposing passwords like that.

I would propose to leave out username and password entirely and instead use --netrc. The user that runs the script should make sure the netrc file is there. And it should be documented in the help text, as in my suggestion above.

tomereli replied:
Good idea, changed to use .netrc

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+                files=$(find "$(dirname "$local_path")/$dir/" -type f -maxdepth 1 -print0 | tr '\0' ',' | sed 's/,$//')
+                dbg "$files"
+                [ -n "$files" ] && {
+                     curl -u "$user":"$password" -T "{$files}" "$OWNCLOUD_URL/$user/$remote_path/$dir/" "$REDIRECT"
```
If you quote $REDIRECT, it will not work. I know shellcheck requests it, but shellcheck is wrong here :-).

arnout commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,76 @@
+                     curl -u "$user":"$password" -T "{$files}" "$OWNCLOUD_URL/$user/$remote_path/$dir/" "$REDIRECT"
+                }
+            done
+            echo # show progress
```
Indentation is wrong here - this should be at the level of the `find`.

arnout reviewed (changes requested):
I only reviewed the first commit but since I have so much feedback and you're actively working on it, I'm already requesting changes.

I'll change it to approve later.

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -22,7 +22,10 @@ usage() {
     echo "      -o|--log-folder - path to put the logs to (make sure it does not contain previous logs)."
     echo "      -e|--easymesh-cert - path to easymesh_cert repository (default ../easymesh_cert)"
-    echo "      -u|--upload - upload results"
+    echo "      --owncloud-upload - weather or not to upload results to owncloud"
```
```suggestion
    echo "      --owncloud-upload - wether or not to upload results to owncloud"
```

tomereli replied:
Its actually `whether`, fixed.

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -22,7 +22,10 @@ usage() {
-    echo "      -u|--upload - upload results"
+    echo "      --owncloud-upload - weather or not to upload results to owncloud"
+    echo "      --owncloud-user - owncloud username"
+    echo "      --owncloud-password - owncloud password"
```
Again, I'd rather rely on .netrc.

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -22,7 +22,10 @@ usage() {
+    echo "      --owncloud-upload - weather or not to upload results to owncloud"
+    echo "      --owncloud-user - owncloud username"
+    echo "      --owncloud-password - owncloud password"
+    echo "      --owncloud-path - relative path in the owncloud server to upload results to"
```
Please mention the default. Note that you can use 
```suggestion
    echo "      --owncloud-path - relative path in the owncloud server to upload results to"
    echo "                                   (default: $OWNCLOUD_PATH)"
```

rmelotte commented on ci/run_easymesh_cert.sh:
```diff
@@ -80,7 +80,7 @@ main() {
 
     info "Start running tests"
-    run "$EASYMESH_CERT_PATH"/run_test_file.sh -o "$LOG_FOLDER" -d "$TARGET_DEVICE" "$TESTS" "$VERBOSE_OPT"
+    "$EASYMESH_CERT_PATH"/run_test_file.sh -o "$LOG_FOLDER" -d "$TARGET_DEVICE" "$TESTS" "$VERBOSE_OPT"
```
I'm not sure quoting "$TESTS" will work if the intention is to pass multiple tests. Did you try it?
I didn't see the rest of the file, but if run_easymesg_cert is just a wrapper to run_test_file.sh then you should be able to use "$@" instead.

tomereli replied:
Yes it works:
```
Done
run_easymesh_cert.sh: Start running tests
The following test(s) will be run:

MAP-4.2.1 MAP-4.3.2_ETH
Applying radio info configuration
Using the following MasterTestInfo.xml: /home/tester/work/easymesh_cert/config/mastertestinfo/default.xml

Running test MAP-4.2.1 (1/2)

Time: 2020-03-27_15-30-29

Applying configuration for an agent tes
```

This was a suggestion by `shellcheck`. Originally I used `"$@"`, and this was the error:
```
Assigning an array to a string! Assign as array, or use * instead of @ to concatenate.shellcheck(SC2124)
```

arnout replied:
Wow... This is an artefact from how run_test_file.sh is written.

As Raphaél observed, the tests will be passed a single argument to run_test_file.sh.

The read_test_names parses that into one line per argument. So all the explicitly specified tests are on a single line. You can see that in the output above: normally, the "The following test(s) will be run" output should print one test per line.

But then, the _run_tests function uses `wc -w` to get the number of tests, and `for t in $test_names` (unquoted!) to iterate over them. So for both cases, word splitting is done again, and it magically works.

It still is a horrible hack though.

The proper solution would be to switch to bash and user arrays.

tomereli replied:
Lets do this in a separate PR then and enjoy this moment of bliss that something magically works :)

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -75,19 +81,12 @@ main() {
-    passed=$(find "$LOG_FOLDER"/PASS/ -maxdepth 1 -type d -printf x | wc -c)
-    failed=$(find "$LOG_FOLDER"/FAIL/ -maxdepth 1 -type d -printf x | wc -c)
-    total=$((failed+passed))
-    if [ "$VERBOSE" = "true" ]; then
-        info "Passed tests:"
-        ls "$LOG_FOLDER/PASS"
-        info "Failed tests:"
-        ls "$LOG_FOLDER/FAIL"
```
This bit is *not* done in run_test_file.sh, and I think it's useful to do it.

rmelotte replied:
Actually I've been wanting to at least print the list of failing tests at the end of the job for some times now, I think it would be useful to do it from run_test_file.sh.

arnout commented on tools/download_ipk.sh:
```diff
@@ -23,7 +23,7 @@ download() {
     # curl options - fail on error (-f), follow redirect (-L)
     curl -f -L "$IPK_URL" --output "$IPK_PATH"/prplmesh.ipk
-    curl -f -L "$BUILDINFO_URL" --output "$IPK_PATH"/prplmesh.buildinfo
+    curl -f -L "$BUILDINFO_URL" --output "$IPK_PATH"/prplmesh_buildinfo.txt
```
Good idea!

arnout commented on tools/download_ipk.sh:
```diff
@@ -22,12 +23,12 @@ download() {
     # curl options - fail on error (-f), follow redirect (-L)
-    curl -f -L "$IPK_URL" --output "$IPK_PATH"/prplmesh.ipk
-    curl -f -L "$BUILDINFO_URL" --output "$IPK_PATH"/prplmesh_buildinfo.txt
+    curl -f -L "$IPK_URL" --output "$IPK_PATH"/prplmesh.ipk "$REDIRECT"
```
As mentioned elsewhere, the non-verbose redirect will not work due to the quotes.

However, specfically for curl, you should use `-s` instead of redirecting output. Also, you probably *do* want to keep errors, so it's better to use `-s -S`.
```suggestion
    curl ${VERBOSE:+ -s -S} -f -L "$IPK_URL" --output "$IPK_PATH"/prplmesh.ipk
```

Note the optional argument handling suggestion from [SC2086](https://github.com/koalaman/shellcheck/wiki/SC2086). Note that this means VERBOSE needs to be defined as empty instead of false by default, so if you want to apply this trick in any script that uses functions.sh, it's a bigger change.

arnout commented on tools/download_ipk.sh:
```diff
@@ -41,6 +42,9 @@ main() {
                 exit 0
                 ;;
+            -v|--verbose)
+                REDIRECT=""
```
```suggestion
                REDIRECT=""
                shift
```

Something tells me you didn't really test this feature :-)

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -71,7 +71,7 @@ main() {
 
     info "download latest ipk from branch $BRANCH"
-    run "$TOOLS_PATH"/download_ipk.sh --branch "$BRANCH"
+    run "$TOOLS_PATH"/download_ipk.sh --quite --branch "$BRANCH" "$VERBOSE_OPT"
```
quite -> quiet - but that option is not actually implemented in download_ipk.sh.

VERBOSE_OPT shouldn't be quoted, otherwise you're passing an extra empty argument when verbose is not set. So with shellcheck's suggestion:
```suggestion
    run "$TOOLS_PATH"/download_ipk.sh --branch "$BRANCH" ${VERBOSE:+ -v}
```

arnout reviewed (approved):
I wouldn't call them nitpicks, but nothing I need to re-review.

arnout commented:
Oh, one more thing: you have commit summary lines like: `ci: run_easymesh_cert.sh add upload support`. For me, that doesn't read very comfortably because it sounds like `run_easymesh_cert.sh` is the subject of the sentence, which means that `add` should be `adds`. So please add an additional colon: `ci: run_easymesh_cert.sh: add upload support`

Also, there's a typo in one of the commit messages: exists -> exits.

Commit:
tools: download_ipk.sh use -s option for curl

Add -s option to the curl command so the progress bar will not be shown,
since it just clatters the log and looks bad in the cron job email.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert.sh exit and print on errors

Add explicit prints and exit on errors instead of using `run` function
which only exits.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert.sh: fix verbose option usage

Pass verbose option as explained in
https://github.com/koalaman/shellcheck/wiki/SC2086.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

rmelotte commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,80 @@
+            while read -r -s dir; do
+                printf . # show progress
+                dbg "Create directory: $dir" 
+                curl ${QUIET:+ -s -S} -f -n -X MKCOL "$OWNCLOUD_URL/$user/$remote_path/$dir"
```
Can you exit the script with an exit code different than 0 if any of these fail? (not necessarily at the first error, but it would be great to know that something went wrong)
It would allow the job to fail in CI when there is an issue with the upload: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/488726605

tomereli replied:
Done

rmelotte commented on ci/upload_to_owncloud.sh:
```diff
@@ -0,0 +1,80 @@
+}
+
+OWNCLOUD_URL="https://ftp.essensium.com/owncloud/remote.php/dav/files"
+OWNCLOUD_BROWSE_URL="https://ftp.essensium.com/owncloud/index.php/apps/files/?dir=/prplmesh/certification"
```
Is this only informational for the user?
It won't always upload to `prplmesh/certification` right?

tomereli replied:
Of course not, this is the base URL for user to be able to click on the link, since the URL used by `curl` can't be used for that.

See here how it is used:
```
info "upload $local_path to $OWNCLOUD_BROWSE_URL/$remote_path/$(basename "$local_path") (using user $user)"
```

arnout replied:
Ah, but that URL is not usable for most people. Instead, you should use
```suggestion
OWNCLOUD_BROWSE_URL="https://ftp.essensium.com/owncloud/index.php/s/ketl3eFhIgwcweZ?path=" (assuming `remote_path` starts with `/Nightly`)
```
